### PR TITLE
Introduce Predicate.t and allow for globs/predicates to be memoized

### PR DESCRIPTION
**Pull request by rgrinberg**
This is another step towards having proper glob dependencies in dune. As a requisite for using the memoization system to calculate globs, we need globs to be comparable for equality. To achieve this, I borrow @diml's idea of providing a user input token that would identify a particular glob.

There are some complications in dune of course, because we actually allow for arbitrary predicates for "globbing", and not all "globs" come from user input. I introduce a `Predicate.t` module to make all of this more explicit.

Also, I cleanup the rest of the codebase not to use raw regular expressions when actually wanting to use globs.

### CODE REVIEW

**Comment by aalekseyev** [comment]
I think we should try to avoid relying so much on programmer discipline and instead make modules robust to abuse. `Predicate.t` introduced here is very easy to abuse by accident and end up using the same identifier for different predicates.

I think at the minimum we should try to:

1.
- make a global registry of identifiers used so far, and raise an exception if the same identifier is used again
- remove `contramap` (or make it take a new identifier string)

Alternatively, we could:
2. Make Predicate a variant type that supports hashing and equality in a natural way. If it needs to be an open variant, we could try something fancy (expose a functor that adds new constructors and requires them to be hashable?), but maybe a closed variant can be good enough at the expense of expressivity?

Or
3. We could have predicates carry meaningless unique int ids allocated in [create] and use those for comparison (thereby making it equivalent to physical equality).

I think the third alternative is actually the best in terms of vallue-for-money. Any reason we shouldn't just do that? Obviously, you lose sharing between two separately allocated predicates that happen to be extensionally equal, but, well, 
- in (1) we'd have to raise in this case, anyway?
- you could recover the physical equality by another layer of memoization (you could memoize the individual functions that create Predicate.t)

**Comment by rgrinberg** [comment]
3. is good but it doesn't seem to work for predicates created from globs. I think this issue complicates 1 as well. Globs already have a good representation as strings so they don't need any identifiers. We need to preserve that when converting them to predicates.

Another possibility is to get rid of `Predicate.t` and use `Predicate_lang.t` anyway. There we can implement faithful comparison based on the AST. But it seems a bit kludgy to define all predicates via this user DSL.

**Comment by aalekseyev** [comment]
I guess "globs are already good" point could be addressed directly, like this:

```
type Predicate.t = 
  | Glob of Glob.t
  | Non_glob of { id : int; f : path -> bool }
```

What do you think?

>  it seems a bit kludgy to define all predicates via this user DSL.

Agreed. Although it doesn't seem crazy either.

**Comment by diml** [comment]
I can't remember, where do we use a function rather than a glob already? Is it just for the bootstrap?

**Comment by aalekseyev** [comment]
FWIW, I believe (3) should still work for predicates created from globs even if you don't do anything special if you care to do the extra layer of memoization I mentioned. Maybe I'm missing something, or maybe you're saying that *without* that layer it's broken.

**Comment by rgrinberg** [comment]
Note that we use both `string Predicate.t` and `Path.t Predicate.t`. I guess we could change the definition to:

```
type _ t = 
  | Glob : Glob.t -> string t
  | Non_glob of { id : int; f : path -> bool } -> Path.t t
```

> I can't remember, where do we use a function rather than a glob already? Is it just for the bootstrap?

To skip js targets for `@check`.

**Comment by diml** [comment]
The function case is required for `Build.path_matching`. Looking at the code, the two use cases for `Build.path_maching` could be encoded as globs. So if the function case is an issue, we can simply remove `Build.patch_matching` and rely on globs only. For the bootstrap, we can have a micro implementation that only supports a few globs.

**Comment by rgrinberg** [comment]
> FWIW, I believe (3) should still work for predicates created from globs even if you don't do anything special if you care to do the extra layer of memoization I mentioned. Maybe I'm missing something, or maybe you're saying that without that layer it's broken.

Yeah, I guess that would work. One thing that makes me suspicious of generate the unique id's is that I'm not entirely sure all these rule generating functions are called in a deterministic order between runs (depending on how you invoke `$ dune build`). But that's a problem if we only start storing them. So if you and Jeremie are alright with this, we can do 3).

**Comment by rgrinberg** [comment]
@diml the predicate to skip the js targets looks like this:

```
(fun p -> not (List.exists js_targets ~f:(Path.equal p)))
```

How would that be encoded as a glob?

**Comment by diml** [comment]
Hmm, well once we have the explicit js mode, we no longer need that, right?

**Comment by rgrinberg** [comment]
If `explicit_js_mode` was the default, I would agree. But now this basically breaks `@all` for everyone that doesn't use it. I expect we'd get a lot of bug reports regarding this.

But in any case, I do think that just having the function case simplifies things for us. If we use `Predicate.t` uniformly, then we can define our predicates in 3 ways:

* Using globs
* Using the predicate DSL
* Directly writing the function

with minimal friction. So i'm still in favor of Arseniy's third option over dropping it altogether.

**Comment by diml** [comment]
Ok. Yh that seems fine to me.

**Comment by rgrinberg** [comment]
@aalekseyev How about this type:

```
module Id = struct
  type t =
    | Generated of int
    | User of Sexp.t
end

type 'a t =
  { id : Id.t
  ; f : 'a -> bool
  }
```

Where the `Generated` id's will be incremented. Globs will be created using the `User` constructor. `contramap` will create a `Generated` id always.

**Comment by aalekseyev** [comment]
I think that should work.

**Comment by rgrinberg** [comment]
I implemented that but there's still a bit of an issue with contramap. We'd like to turn a `string Predicate.t` into a `Path.t Predicate.t` using `Path.basename`, but using `contramap` means that our glob identifier is discarded.

I see a couple of ways around this:

* We stick to using `string Predicate.t` in `build_system.ml`. Not very attractive but will work.

* We make contramap preserve the id's but mark it as "not safe" somehow. It should work fine as long as `f` is pure and doesn't capture anything from the environment. I don't think that's so bad since you basically have to worry about that when writing memoized code anyway.

* We introduce `basename` as a safe primitive. E.g. `val on_basename : string Predicate.t -> Path.t Predicate.t`.

@aalekseyev any preferences? (or better suggestions perhaps).

I don't think it's so important to get this perfect from the first try by the way. The end goal here is to make globs first class. The details can be rehashed later.

**Comment by diml** [comment]
If we simply change the signature of `Build.paths_matching` to take a `(string -> bool)` function rather than a `(Path.t -> bool)`, does that avoid the isue altogether?

**Comment by rgrinberg** [comment]
@diml yes it does (I think that corresponds to my 1st) suggestion. It will make the defining the `Path.equal` based predicate more awkward but it's no big deal

**Comment by diml** [comment]
If it avoids these ids considerations, let's do that.

**Comment by diml** [comment]
And in fact, we can completely get rid of functions here: `(fun p -> not (List.exists js_targets ~f:(Path.equal p)))` can be written as `(not (or js_target1 js_target2 ...))`.

**Comment by rgrinberg** [comment]
> If it avoids these ids considerations, let's do that.

It does. Okay let's do that and just write all predicates via globs and the plang.

> And in fact, we can completely get rid of functions here: (fun p -> not (List.exists js_targets ~f:(Path.equal p))) can be written as (not (or js_target1 js_target2 ...)).

Yeah, that would work as well. Note that we'll have to mess around with basenames here. The alternative would be to generalize `plang` a bit to work on arbitrary predicates.

**Comment by diml** [comment]
> Yeah, that would work as well. Note that we'll have to mess around with basenames here.

That seems fine to me. Should we do that and remove the `Predicate` module then?

**Comment by aalekseyev** [comment]
Even though expressiveness reduction works, I want to point out that "memoize and then use physical equality" trick discussed before is pretty powerful and it would have worked for basenames or contramap or what have you.
Something along these lines (modulo the memoization boilerplate):

let contramap_basename = memoize ... (contramap ~f:basename)

should have worked.

**Comment by diml** [comment]
Is it specific to predicates though? I'm not against it but I just to make sure we are introducing the right concepts.

**Comment by rgrinberg** [comment]
Making it work via `Predicate_lang.t` is also a bit annoying. We have to write predicates like this:

```
    Predicate_lang.of_pred (function
      | "" -> false
      | s -> s.[0] <> '.' && s.[0] <> '_')
```

Directly via the predicate language. You have to write something like: `[^._] \ ""`.

Just to re-iterate the issues:

* It is useful to be able to write predicates as globs or functions. But we can do without the latter with a bit of pain

* I think it will be very useful to persist predicates with the memoizing build system. If we allow for predicates like `**/*.v` for example. These are not so cheap to evaluate. At the same time, the majority of predicates are fine to re-evaluate.

**Comment by diml** [comment]
Yh, I'm basically OK with that in the end, but I'm not sure the module should be called `Predicate` and be only about predicates. It seems to me that this is a more general concept that is not specific to predicates. What about `Cell` or `Phys`?

**Comment by rgrinberg** [comment]
The name phys implies that we'll be using physical equality. That will not work with persisting the globs. Am I correct in assuming that? If so then I think we should avoid that name

**Comment by aalekseyev** [comment]
Yeah, the thing Jeremie is talking about is just the trick of using a generated id + physical equality for memoization keys.

Indeed to achieve persistence we should avoid that. If we want to have both persisted and non-persisted keys, we could have two different memo's for them, or perhaps we could introduce a "mixed-persistence" memo, where only some of the keys are persisted...

File: src/predicate.mli [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+type 'a t
+
+val equal : 'a t -> 'a t -> bool
+
+val hash : _ t -> int
+
+val to_sexp : _ t Sexp.Encoder.t
+
+val create : id:Sexp.t Lazy.t -> f:('a -> bool) -> 'a t
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/predicate.mli]: done

**Pull request merged**